platform/linuxbsd/wayland/display_server_wayland.cpp

```diff
@@ -0,0 +1,1353 @@
+	MutexLock mutex_lock(wayland_thread.mutex);
+
+	Key key = wayland_thread.keyboard_get_key_from_physical(p_keycode);
+
+	// If not found, fallback to QWERTY.
+	// This should match the behavior of the event pump.
+	if (key == Key::NONE) {
+		return p_keycode;
+	}
+
+	if (key >= Key::A + 32 && key <= Key::Z + 32) {
+		key -= 'a' - 'A';
+	}
+
+	// Make it consistent with the keys returned by `Input`.
+	if (key == Key::BACKTAB) {
+		key = Key::TAB;
+	}
+
+	return key;
+}
+
+void DisplayServerWayland::process_events() {
+	wayland_thread.mutex.lock();
+
+	while (wayland_thread.has_message()) {
+		Ref<WaylandThread::Message> msg = wayland_thread.pop_message();
+
+		Ref<WaylandThread::WindowRectMessage> winrect_msg = msg;
+		if (winrect_msg.is_valid()) {
+			_resize_window(winrect_msg->rect.size);
+		}
+
+		Ref<WaylandThread::WindowEventMessage> winev_msg = msg;
+		if (winev_msg.is_valid()) {
+			_send_window_event(winev_msg->event);
+
+			if (winev_msg->event == WINDOW_EVENT_FOCUS_IN) {
+				if (OS::get_singleton()->get_main_loop()) {
+					OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_APPLICATION_FOCUS_IN);
+				}
+			} else if (winev_msg->event == WINDOW_EVENT_FOCUS_OUT) {
+				if (OS::get_singleton()->get_main_loop()) {
+					OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_APPLICATION_FOCUS_OUT);
+				}
+			}
+		}
+
+		Ref<WaylandThread::InputEventMessage> inputev_msg = msg;
+		if (inputev_msg.is_valid()) {
+			Input::get_singleton()->parse_input_event(inputev_msg->event);
+		}
+
+		Ref<WaylandThread::DropFilesEventMessage> dropfiles_msg = msg;
+		if (dropfiles_msg.is_valid()) {
+			WindowData wd = main_window;
+
+			if (wd.drop_files_callback.is_valid()) {
+				wd.drop_files_callback.call(dropfiles_msg->files);
+			}
+		}
+	}
+
+	wayland_thread.keyboard_echo_keys();
+
+	frame = wayland_thread.get_reset_frame();
+
+	wayland_thread.mutex.unlock();
+
+	Input::get_singleton()->flush_buffered_events();
+}
+
+void DisplayServerWayland::release_rendering_thread() {
+#ifdef GLES3_ENABLED
+	if (egl_manager) {
+		egl_manager->release_current();
+	}
+#endif
+}
+
+void DisplayServerWayland::make_rendering_thread() {
+#ifdef GLES3_ENABLED
+	if (egl_manager) {
+		egl_manager->make_current();
+	}
+#endif
+}
+
+void DisplayServerWayland::swap_buffers() {
+#ifdef GLES3_ENABLED
+	if (egl_manager) {
+		egl_manager->swap_buffers();
+	}
+#endif
+}
+
+void DisplayServerWayland::set_context(Context p_context) {
+	MutexLock mutex_lock(wayland_thread.mutex);
+
+	DEBUG_LOG_WAYLAND(vformat("Setting context %d.", p_context));
+
+	context = p_context;
+
+	String app_id = _get_app_id_from_context(p_context);
+	wayland_thread.window_set_app_id(MAIN_WINDOW_ID, app_id);
+}
+
+Vector<String> DisplayServerWayland::get_rendering_drivers_func() {
+	Vector<String> drivers;
+
+#ifdef VULKAN_ENABLED
+	drivers.push_back("vulkan");
+#endif
+
+#ifdef GLES3_ENABLED
+	drivers.push_back("opengl3");
```

Comment:
not a big deal i think but FWIW, `display_server_x11.cpp` calls this same driver `opengl3_es`, `opengl3` is the desktop GL

Reply:
Oh, thought wayland used EGL+GLES, that's fine then!